hermes/probe/metric.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: Evan Spendlove, GitHub: evanSpendlove.
+//
+// Metric implements the latency distribution metrics for Hermes.
+// Package probe implements the probe that Hermes uses to monitor
+// a storage system.
+
+package probe
+
+import (
+	"fmt"
+	"github.com/google/cloudprober/metrics"
+	"time"
+
+	metricpb "github.com/google/cloudprober/metrics/proto"
+	probepb "github.com/googleinterns/step224-2020/config/proto"
+)
+
+// probeRunMetrics stores the cumulative metrics for probe runs for a target.
+// These metrics are sent via a data channel at the end of the probe run.
+type probeRunMetrics struct {
+	StorageSystem string
+	Target        string
+	BucketName    string
+
+	// probeOpLatency is used to record latency values with distinct labels
+	// per exit status per probe operation per target.
+	// Recommended usage: probeOpLatency[probeOp][exitStatus].Metric("latency").AddFloat64(<val>)
+	ProbeOpLatency map[string]map[string]*metrics.EventMetrics
+	// apiCallLatency is used to record latency values with distinct labels
+	// per exit status per API call per target.
+	// Recommended usage: apiCallLatency[apiCall][exitStatus].Metric("latency").AddFloat64(<val>)
+	ApiCallLatency map[string]map[string]*metrics.EventMetrics
+}
+
+// NewProbeMetrics creates a new *probeRunMetrics object and initialises the fields inside it.
+// Arguments:
+//	- conf: pass a HermesProbeDef config
+//	- target: pass the target for which metrics are to be collected.
+// Returns:
+//	- pm: returns an initialised *probeRunMetrics object.
+//	- err: returns an error if a latency distribution cannot be created from the config proto.
+func NewProbeMetrics(conf probepb.HermesProbeDef, target probepb.Target) (*probeRunMetrics, error) {
+	probeOps := []string{"total_probe_run", "check_nil", "read_and_verify", "delete_file", "create_file"}
+	apiCalls := []string{"list_files", "create_file", "delete_file", "get_file"}
+	probeExitStatuses := []string{"success", "op_timeout", "probe_failed", "file_missing", "bucket_missing", "file_corrupted", "file_read_failure", "file_metadata_mismatch", "unknown_file_found", "all_files_missing"}
```

Comment:
Ideally  put all these metric things into a var, in a seperate file, or at least var block at the top - this is likely to be the bit that people look for the most in the file, so it should be easy to find.

Reply:
I've moved these into a series of consts and maps.